src/main/kotlin/service/user/Registration.kt

```diff
@@ -0,0 +1,48 @@
+package service.user
+
+import model.Users
+import org.jetbrains.exposed.sql.insert
+import org.jetbrains.exposed.sql.select
+import org.jetbrains.exposed.sql.transactions.transaction
+import org.mindrot.jbcrypt.BCrypt
+
+object Registration {
+    fun registerUser(registrationRequest: UserRegistrationRequest): UserAuthenticationResponse {
+        if (userAlreadyExists(registrationRequest.username)) {
+            return UserAuthenticationResponse(error = "A user with your chosen name already exists")
+        } else if (registrationRequest.password.length < 6) {
+            return UserAuthenticationResponse(error = "Password should be at least 6 characters long")
+        } else if (registrationRequest.username.length < 4) {
+            return UserAuthenticationResponse(error = "Username should be at least 4 characters long")
+        } else if (!userNameIsValid(registrationRequest.username)) {
+            return UserAuthenticationResponse(error = "Login should be consists of digits, letters, dots or underscores")
+        } else {
+            val hashedPassword = BCrypt.hashpw(registrationRequest.password, BCrypt.gensalt())
+
+            return try {
+                transaction {
+                    Users.insert {
+                        it[name] = registrationRequest.username
+                        it[email] = registrationRequest.email
+                        it[password] = hashedPassword
+                    }
+                }
+
+                val userToken = TokenManager.createToken(registrationRequest.username)
+
+                UserAuthenticationResponse(ok = true, token = userToken)
+            } catch (exception: Exception) {
+                UserAuthenticationResponse(error = "An unidentified error occurred during the account creation")
+            }
+        }
+    }
+
+    private fun userAlreadyExists(username: String): Boolean {
+        val userAlreadyExists = transaction { Users.select { Users.name eq username }.firstOrNull() }
+        return userAlreadyExists != null
+    }
+
+    private fun userNameIsValid(username: String): Boolean {
+        return Regex("^[\\pL\\p{Mn}\\p{Nd}\\p{Pc}]+$").matches(username)
```

Comment:
So what does this filter? Is it purely database sanity? Or do we also don't want certain odd things to show on the frontend. I'd love to have an emoji in my username, am I allowed to?

Reply:
Just kidding about the emoji btw.

Reply:
It checks if the username only contains letters, digits, or punctuating characters.
@CJStuart Could you maybe explain the regex further?

Reply:
The `\p{...}` classes are unicode character classes. Basically it includes all letters, decimal numbers, accents, and connecting punctuation (e.g. _). It sadly does not include the `\p{So}` class, for emoji and other symbols, though. 

These are all the categories: http://www.fileformat.info/info/unicode/category/index.htm